libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -942,3 +942,89 @@ void test_MQTT_Connect_Restore_Session( void )
+     * QoS 2 PUBLISH transaction in this connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, subscribeToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS2 ) );
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_TRUE( receivedSubAck );
+
+    /* Publish to the same topic with Qos 2 (so that the broker can re-publish it back to us). */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS2 ) );
+
+    /* Disconnect on receiving PUBREL so that we are not able to complete in the incoming QoS2
+     * PUBLISH in the current connection. */
+    disconnectOnPacketType = MQTT_PACKET_TYPE_PUBREL;
+    TEST_ASSERT_EQUAL( MQTTSendFailed,
+                       MQTT_ProcessLoop( &context, 3 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Verify that the connection with the broker has been disconnected. */
+    TEST_ASSERT_EQUAL( MQTTNotConnected, context.connectStatus );
+
+    /* We will re-establish an MQTT over TLS connection with the broker to restore the persistent session. */
+
+    /* Create a new TLS+TCP network connection with the server. */
+    TEST_ASSERT_EQUAL( OPENSSL_SUCCESS, Openssl_Connect( &networkContext,
+                                                         &serverInfo,
+                                                         &opensslCredentials,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS ) );
+    TEST_ASSERT_NOT_EQUAL( -1, networkContext.socketDescriptor );
+    TEST_ASSERT_NOT_NULL( networkContext.pSsl );
+
+    /* Re-establish the persistent session with the broker by connecting with "clean session" flag set to 0. */
+    establishMqttSession( &context, &networkContext, false, &persistentSession );
+
+    /* Verify that the session was resumed. */
+    TEST_ASSERT_TRUE( persistentSession );
+
+    /* Clear the global variable for not disconnecting on PUBREL
+     * that we receive from the broker on the session restoration. */
+    disconnectOnPacketType = MQTT_PACKET_TYPE_INVALID;
+
+    /* Resume the incomplete incoming QoS 2 PUBLISH transaction from the previous MQTT connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Make sure that the broker resent the PUBREL packet on session restoration. */
+    TEST_ASSERT_TRUE( receivedPubRel );
+
+    /* Make sure that the library sent a PUBCOMP packet in response to the PUBREL packet
+     * from the server to complete the incoming PUBLISH QoS2 transaction. */
+    TEST_ASSERT_EQUAL( MQTTPubCompSend, context.incomingPublishRecords[ 0 ].publishState );
+    TEST_ASSERT_EQUAL( 0u, context.incomingPublishRecords[ 0 ].packetId );
+    TEST_ASSERT_EQUAL( true, context.controlPacketSent );
```

Comment:
Why not just `TEST_ASSERT_TRUE( context.controlPacketSent );`?

Reply:
`true` is platform dependent..

Reply:
I don't understand. You used `TEST_ASSERT_TRUE` 5 lines above this. Why not here?